Sources/Build/BuildDelegate.swift

```diff
@@ -684,3 +689,66 @@ private extension Diagnostic.Message {
         .error("failed parsing the Swift compiler output: \(error)")
     }
 }
+
+#if canImport(CryptoKit)
+import CryptoKit
+
+/// A filesystem for LLBuild that returns hash of the files
+/// instead of stat information. This is obviously a hack
+/// until llbuild has proper support for content-based tracking.
+fileprivate struct LLBContentHashFileSystem: SPMLLBuild.FileSystem {
+    public struct HashFileInfo: SPMLLBuild.FileInfo {
+        var statBuf: stat
+
+        init(_ statBuf: stat) {
+            self.statBuf = statBuf
+        }
+    }
+
+    func fileContents(_ path: String) throws -> ByteString {
+        try localFileSystem.readFileContents(AbsolutePath(path))
+    }
+
+    func read(_ path: String) throws -> [UInt8] {
+        try fileContents(path).contents
+    }
+
+    func getFileInfo(_ path: String) throws -> SPMLLBuild.FileInfo {
+        if #available(macOS 10.15, *) {
+            var hasher = CryptoKit.SHA256()
+            hasher.update(data: try read(path))
+
+            var hash = [UInt8]()
+            hasher.finalize().withUnsafeBytes { pointer in
+                hash.append(contentsOf: pointer)
+            }
+
+            // Mangle the hash into some of the fields used by llbuild.
+            var statbuf = stat()
+            hash[0..<8].withUnsafeBytes {
+                let ptr = $0.bindMemory(to: Int64.self)
```

Comment:
Is there any risk of alignment issues here, i.e. are the bytes of `hash` guaranteed to be Int64-aligned on all possible platforms and architectures?

Reply:
Yeah, this will likely crash on ARM.